File: src/ast.c
#include <stdio.h>
#include <assert.h>

#include "object.h"
#include "number.h"
#include "message.h"
#include "skstring.h"
#include "list.h"
#include "block.h"
#include "ast.h"

SkObject *sk_ast_evaluate_block(SkVM *vm, SkNode *node) {
    SkObject *self = sk_block_clone(sk_vm_get_proto(vm, "Block"));
    /* set the initial value */
    sk_block_set_block(self, (SkNodeBlock *)node);
    return self;
}

SkObject *sk_ast_evaluate_message(SkVM *vm, SkNode *node_) {
    SkNodeMessage *node = (SkNodeMessage *)node_;
    SkObject *self = sk_message_clone(sk_vm_get_proto(vm, "Message"));
    sk_message_set_name(self, sk_ast_evaluate_string(vm, node->name));
    SkObject *receiver = sk_ast_evaluate(vm, node->receiver);
    if(receiver == vm->nil) {
        receiver = vm->lobby;
    }
    sk_message_set_receiver(self, receiver);

    SkObject *args = sk_message_clone(sk_vm_get_proto(vm, "List"));
    int i;
    for(i = 0; i < node->argcount; i++) {
        SkObject *argument = sk_ast_evaluate_message(vm, node->arguments[i]);
        sk_list_append(args, argument);
    }

    return sk_object_dispatch_message(receiver, self);
}

SkObject *sk_ast_evaluate_string(SkVM *vm, SkNode *node) {
    SkObject *self = sk_string_clone(sk_vm_get_proto(vm, "String"));
    /* set the initial value */
    sk_string_set_bstring(self, ((SkNodeString *)node)->value);
    return self;
}

/* clone and return a number */
SkObject *sk_ast_evaluate_number(SkVM *vm, SkNode *node) {
    SkObject *self = sk_number_clone(sk_vm_get_proto(vm, "Number")); 
    /* set the initial value */
    sk_number_set_int(self, ((SkNodeNumber *)node)->value);
    return self;
}

SkObject *sk_ast_evaluate(SkVM *vm, SkNode *node) {
    switch(node->type) {
        case SK_NODE_TYPE_NUMBER:
            return sk_ast_evaluate_number(vm, node);
        case SK_NODE_TYPE_BLOCK:
            return sk_ast_evaluate_block(vm, node);
        case SK_NODE_TYPE_STRING:
            return sk_ast_evaluate_string(vm, node);
        case SK_NODE_TYPE_MESSAGE:
            return sk_ast_evaluate_message(vm, node);
        case SK_NODE_TYPE_NIL:
            return vm->nil;
        default:
            printf("Error: " __FILE__ ": No idea what node type %d is.\n", node->type);
            return NULL;
    }
}

File: pyskelde/ast.py
from struct import pack

from . import opcodes as op

class Node(object):
    def __init__(self):
        pass

    def build(self):
        raise NotImplementedError()

class String(Node):
    def __init__(self, value):
        Node.__init__(self)
        self.value = value

    def build(self):
        return pack('ii%dc' % len(self.value), op.STRING, len(self.value), *self.value)

class Number(Node):
    def __init__(self, value):
        Node.__init__(self)
        self.value = value

    def build(self):
        return pack('ii', op.NUMBER, self.value)

class Block(Node):
    def __init__(self, children=None):
        Node.__init__(self)
        if children is None:
            children = []
        self.children = []

    def build(self):
        code = pack('ii', op.BLOCK, len(self.children))
        for child in self.children:
            code += child.build()
        return code

class Nil(Node):
    def build(self):
        return pack('i', op.NIL)

NIL = Nil()

class Message(Node):
    def __init__(self, name, receiver=NIL, arguments=None):
        """
            :Parameters:
                `name` : str or String
                    Name of the message
        """
        Node.__init__(self)
        self.name = name
        self.receiver = receiver
        if arguments is None:
            arguments = []
        self.arguments = arguments

    def build(self):
        code = pack('i', op.MESSAGE)
        name = self.name
        if isinstance(name, basestring):
            name = String(name)
        # name
        code += name.build()
        # receiver
        code += self.receiver.build()
        # arguments
        code += pack('i', len(self.arguments))
        for argument in self.arguments:
            code += argument.build()
        print repr(code)
        return code

